examples/src/main/scala/io/chrisdavenport/shellfish/contacts/core/ContactManager.scala

```diff
@@ -0,0 +1,142 @@
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+package io.chrisdavenport.shellfish.contacts.core
+
+import cats.syntax.all.*
+import cats.effect.IO
+import fs2.io.file.Path
+import io.chrisdavenport.shellfish.syntax.path.*
+import io.chrisdavenport.shellfish.contacts.domain.contact.*
+
+trait ContactManager {
+
+  def addContact(contact: Contact): IO[Username]
+
+  def removeContact(username: Username): IO[Unit]
+
+  def searchId(username: Username): IO[Option[Contact]]
+  def searchName(name: Name): IO[List[Contact]]
+  def searchEmail(email: Email): IO[List[Contact]]
+  def searchNumber(number: PhoneNumber): IO[List[Contact]]
+
+  def getAll: IO[List[Contact]]
+
+  def updateContact(username: String)(modify: Contact => Contact): IO[Contact]
+
+}
+
+object ContactManager {
+  def impl(bookPath: Path): ContactManager = new ContactManager {
+
+    private def parseContact(contact: String): Either[Exception, Contact] =
+      contact.split('|') match {
+        case Array(id, firstName, lastName, phoneNumber, email) =>
+          Contact(id, firstName, lastName, phoneNumber, email).asRight
+
+        case _ => new Exception(s"Invalid contact format: $contact").asLeft
+      }
+
+    private def encodeContact(contact: Contact): String =
+      s"${contact.username}|${contact.firstName}|${contact.lastName}|${contact.phoneNumber}|${contact.email}"
+
+    override def addContact(contact: Contact): IO[Username] = {
+      bookPath.readLines.flatMap { lines =>
+        lines.find(_.startsWith(contact.username)) match {
+          case Some(_) =>
+            ContactFound(contact.username).raiseError[IO, Username]
+
+          case None =>
+            lines.isEmpty
+              .pure[IO]
+              .ifM(
+                bookPath.write(encodeContact(contact)).as(contact.username),
+                bookPath.appendLine(encodeContact(contact)).as(contact.username)
+              )
+        }
+      }
+    }
+
+    override def removeContact(username: Username): IO[Unit] =
+      for {
+        contacts <- bookPath.readLines
+        _ <- bookPath.writeLines(
+          contacts.filterNot(_.startsWith(username))
+        )
+      } yield ()
+
+    override def searchId(username: Username): IO[Option[Contact]] =
+      for {
+        contacts <- bookPath.readLines
+      } yield contacts
+        .find(_.startsWith(username))
+        .flatMap(parseContact(_).toOption)
+
+    override def searchName(name: Name): IO[List[Contact]] =
+      for {
+        contacts <- bookPath.readLines.map(_.filter(_.contains(name)))
```

Comment:
Maybe `bookPath.readLines.map(_.filter(_.contains(name)))` is worth an auxiliary function as you repeated it 3 times?